### PR TITLE
*: Add Watch progress report

### PR DESCRIPTION
**Pull request by gyuho**
For https://github.com/coreos/etcd/issues/4628.

### CODE REVIEW

File: etcdserver/api/v3rpc/watch.go [hunk 1]
```diff
@@ -117,10 +123,30 @@ func (sws *serverWatchStream) recvLoop() error {
...
+			}
+
+			if uv.CreateRequest.ProgressReport {
+				go func() {
+					for {
+						select {
+						case <-sws.reportDone:
+							close(sws.reportStream)
+							return
+						case <-time.After(time.Duration(uv.CreateRequest.ReportInterval) * time.Second):
```
Comment (heyitsanthony) [inline etcdserver/api/v3rpc/watch.go]: Potential race: events posted to watcher channel for rev1, watcher revision increased to rev2, timer fires and progress report is sent with rev2, sendLoop then receives events from watcher channel and sends events for rev1 after reporting progress at rev2. 

This should at least be in the sendLoop.

File: etcdserver/etcdserverpb/rpc.proto [hunk 2]
```diff
@@ -267,6 +267,10 @@ message WatchCreateRequest {
   bytes range_end = 2;
   // start_revision is an optional revision (including) to watch from. No start_revision is "now".
   int64 start_revision = 3;
+  // ProgressReport 'true' subscribes to periodic watch status updates.
+  bool progress_report = 4;
+  // ReportInterval is progress report frequency in seconds. If 5, it notifies every 5 seconds.
```
Comment (xiang90) [inline etcdserver/etcdserverpb/rpc.proto]: report frequency should be at minute level, not second level i think. say we have 100K watchers. 5 second per watch = 20k request / second. also we usually keep hours history, minute level reporting is already more than enough.

File: etcdserver/api/v3rpc/watch.go [hunk 3]
```diff
@@ -117,10 +123,30 @@ func (sws *serverWatchStream) recvLoop() error {
...
+							sws.reportStream <- &pb.WatchResponse{
```
Comment (xiang90) [inline etcdserver/api/v3rpc/watch.go]: basically return https://github.com/coreos/etcd/blob/master/storage/watchable_store.go#L347 in sync way. this is source of truth.

File: etcdserver/etcdserverpb/rpc.proto (hunk 2 above)
Comment (gyuho) [inline etcdserver/etcdserverpb/rpc.proto]: *overwrite inside the function `WithProgressReport(i time.Duration)`

**Comment by gyuho** [comment]
@xiang90 

> i think a better way is to check if the "watcher" (not stream) is currently synced. if it is, then report its rev, which is exact the current storage revision.

Do you mean that check if the watcher id is in synced and report only the synced ones?
Then we will have to add another interface to `WatchStream` I think. Is that ok?

**Comment by xiang90** [comment]
> Do you mean that check if the watcher id is in synced and report only the synced ones?
Then we will have to add another interface to WatchStream I think. Is that ok?

Well... It depends on your implementation. Our goal is to report accurate progress for synced watchers. If it also works for unsynced watcher, then great. If makes it also work for unsynced is complicated, then just ignore unsynced watchers. 

What we do not want to see is to make our impl complicated just to ignore unsynced watchers.

**Comment by gyuho** [comment]
Ok I will try both, and see how it turns out. Thanks!

File: etcdserver/api/v3rpc/watch.go [hunk 4]
```diff
@@ -117,10 +123,33 @@ func (sws *serverWatchStream) recvLoop() error {
...
+
+			if uv.CreateRequest.ProgressReport {
+				go func() {
+					for {
+						select {
+						case <-sws.reportDone:
+							close(sws.reportStream)
+							return
+						case <-time.After(time.Duration(uv.CreateRequest.ReportInterval) * time.Second):
+							if sws.watchStream.WatchStatus(int64(id)) != storage.Synced {
```
Comment (gyuho) [inline etcdserver/api/v3rpc/watch.go]: @xiang90 Just added this method from `storage` package. Can you take a look? Thanks.
(I still have to figure out the fix for race conditions that Anthony pointed out).

**Comment by gyuho** [comment]
After second thought, I think this is too complicated. So we want `watcher.cur` because that's the correct revision that the watch call wants.

How about adding a field `KVRevision` in `WatchResponse` regardless of `synced` or `unsynced`?
So that we don't have to worry about filling in the correct revision in the header?

**Comment by gyuho** [comment]
This needs rework based on our watchable storage implementation (https://github.com/coreos/etcd/pull/4669).

**Pull request closed without merging**